File: firebase/functions/lib/utils/get_server_timestamp.dart
import 'dart:async';

import 'package:firebase_functions_interop/firebase_functions_interop.dart';
import '../on_call_function.dart';
import 'package:data_models/cloud_functions/requests.dart';

class GetServerTimestamp extends OnCallMethod<GetServerTimestampRequest> {
  GetServerTimestamp()
      : super(
          GetServerTimestampRequest.functionName,
          (jsonMap) => GetServerTimestampRequest.fromJson(jsonMap),
        );

  @override
  Future<Map<String, dynamic>> action(
    GetServerTimestampRequest request,
    CallableContext context,
  ) async {
    final result =
        GetServerTimestampResponse(serverTimestamp: DateTime.now().toUtc())
            .toJson();
    print(result);
    return result;
  }

  @override
  void register(FirebaseFunctions functions) {
    functions[functionName] = functions
        .runWith(
          runWithOptions ??
              RuntimeOptions(
                timeoutSeconds: 60,
                memory: '1GB',
                minInstances: 1,
              ),
        )
        .https
        .onCall(callAction);
  }
}

File: firebase/functions/src/utils/get_server_timestamp.ts
import * as functions from 'firebase-functions'
import * as admin from 'firebase-admin'
import { OnCallMethod } from '../on_call_function'
import { GetServerTimestampRequest, GetServerTimestampResponse } from '../types'

export class GetServerTimestamp extends OnCallMethod<GetServerTimestampRequest> {
    constructor() {
        super('serverTimestamp', (data) => (data ?? {}) as GetServerTimestampRequest, {
            runWithOptions: { timeoutSeconds: 60, memory: '1GB', minInstances: 1 },
        })
    }

    async action(
        _request: GetServerTimestampRequest,
        _context: functions.https.CallableContext
    ): Promise<GetServerTimestampResponse> {
        return {
            serverTimestamp: new Date().toISOString(),
        }
    }
}
